packages/core/src/browser/frontend-connection-status.ts

```diff
@@ -0,0 +1,221 @@
+
+        if (this.state === ConnectionState.CONNECTION_LOST) {
+            this.health = 0;
+        } else {
+            this.health = this.history.length === 0 ? 100 : Math.round((this.history.filter(success => success).length / this.history.length) * 100);
+        }
+    }
+
+    next(success: boolean): ConnectionStateMachine {
+        const newHistory = this.updateHistory(success);
+        // Initial optimism.
+        let hasConnection = true;
+        if (newHistory.length > this.props.threshold) {
+            hasConnection = newHistory.slice(-this.props.threshold).some(s => s);
+        }
+        // Ideally, we do not switch back to online if we see any `true` items but, let's say, after three consecutive `true`s.
+        return new ConnectionStateMachine(this.props, hasConnection ? ConnectionState.CONNECTED : ConnectionState.CONNECTION_LOST, newHistory);
+    }
+
+    private updateHistory(success: boolean) {
+        const updated = [...this.history, success];
+        if (updated.length > ConnectionStateMachine.MAX_HISTORY) {
+            updated.shift();
+        }
+        return updated;
+    }
+
+}
+
+@injectable()
+export class ApplicationConnectionStatusContribution implements ConnectionStatusContribution {
+
+    private dialog: ConnectionStatusDialog | undefined;
+    private state = ConnectionState.CONNECTED;
+
+    constructor(
+        @inject(MessageService) protected readonly messageService: MessageService,
+        @inject(ILogger) protected readonly logger: ILogger
+    ) { }
+
+    onStatusChange(event: ConnectionStatusChangeEvent): void {
+        if (this.state !== event.state) {
+            this.state = event.state;
+            switch (event.state) {
+                case ConnectionState.CONNECTION_LOST: {
+                    const message = 'Application connection to the Theia backend is lost. Performing retries...';
+                    this.logger.error(message);
+                    this.messageService.error(message);
+                    this.getOrCreateDialog().open();
+                    break;
+                }
+                case ConnectionState.CONNECTED: {
+                    const message = 'Application connection to the Theia backend was successfully re-established.';
+                    this.logger.info(message);
+                    this.messageService.info(message);
+                    if (this.dialog !== undefined) {
+                        // tslint:disable-next-line:no-any
+                        (<any>this.dialog).accept();
+                    }
+                    break;
+                }
+            }
+        }
+    }
+
+    private getOrCreateDialog() {
+        if (this.dialog === undefined) {
+            this.dialog = new ConnectionStatusDialog();
+        }
+        return this.dialog;
+    }
+
+}
+
+class ConnectionStatusDialog extends AbstractDialog<void> {
+
+    public readonly value: void;
+
+    constructor() {
+        super({ title: 'Not connected' });
```

Comment:
title: `Theia appears to be offline.`

Reply:
one could read this as "Theia server is offline", right? e.g. browsers say something like "there is no connection", if host is not reachable.